File: plateform/frontend/src/types/map.ts
import type { Feature, FeatureCollection, LineString, MultiPolygon, Point } from "geojson";

import type { AssetType, OperationalStatus } from "./assets";
import type { IncidentClassification, IncidentStatus, SeverityTier } from "./incidents";
import type { FreshnessState } from "./telemetry";

export interface ZoneFeatureProperties {
  code: string;
  name: string;
  country: string;
  region: string;
  asset_count: number;
  active_incident_count: number;
}

export interface PipelineFeatureProperties {
  external_id: string;
  name: string;
  status: string;
  criticality: number;
  zone: string;
  population_served: number;
  active_incident_count: number;
  connected_asset_count: number;
}

export interface AssetFeatureProperties {
  external_id: string;
  name: string;
  asset_type: AssetType;
  operational_status: OperationalStatus;
  health_score: number | null;
  zone: string;
  zone_id: string | null;
  location_label: string | null;
  has_cellular_identity: boolean;
  device_msisdn_masked: string | null;
  pipeline_segment: string | null;
  active_incident_count: number;
  last_seen_at: string | null;
  latest_reading_at: string | null;
  telemetry_freshness: FreshnessState | null;
  latest_packet_loss_pct: number | null;
  latest_signal_strength_dbm: number | null;
  latest_battery_pct: number | null;
}

export interface IncidentFeatureProperties {
  incident_number: string;
  title: string;
  severity: SeverityTier;
  classification: IncidentClassification;
  status: IncidentStatus;
  detected_at: string;
  related_asset_id: string | null;
  related_segment_id: string | null;
  summary: string | null;
}

export interface DetectionFeatureProperties {
  detection_number: string;
  priority: string;
  status: string;
  anomaly_score: number;
  trigger_reason: string;
  rule_code: string;
  sensor_id: string;
  zone: string;
  detected_at: string;
  data_mode: string;
}

export type ZoneFeature = Feature<MultiPolygon, ZoneFeatureProperties>;
export type PipelineFeature = Feature<LineString, PipelineFeatureProperties>;
export type AssetFeature = Feature<Point, AssetFeatureProperties>;
export type IncidentFeature = Feature<Point, IncidentFeatureProperties>;
export type DetectionFeature = Feature<Point, DetectionFeatureProperties>;

export type ZoneCollection = FeatureCollection<MultiPolygon, ZoneFeatureProperties>;
export type PipelineCollection = FeatureCollection<LineString, PipelineFeatureProperties>;
export type AssetCollection = FeatureCollection<Point, AssetFeatureProperties>;
export type IncidentCollection = FeatureCollection<Point, IncidentFeatureProperties>;
export type DetectionCollection = FeatureCollection<Point, DetectionFeatureProperties>;

export interface MapSummary {
  visible_zones: number;
  visible_pipelines: number;
  visible_assets: number;
  online: number;
  degraded: number;
  offline: number;
  active_incidents: number;
  critical_incidents: number;
  last_data_update: string | null;
  data_mode: "seeded_demo";
}

export interface NearbyItem {
  feature_type: "asset" | "pipeline" | "incident";
  id: string;
  name: string;
  distance_m: number;
  geometry: Point | LineString | MultiPolygon;
  properties: AssetFeatureProperties | PipelineFeatureProperties | IncidentFeatureProperties;
}

export interface NearbyResponse {
  origin: { latitude: number; longitude: number };
  radius_m: number;
  items: NearbyItem[];
  total: number;
}

export interface MapFilters {
  zone: string;
  asset_type: AssetType | "";
  asset_status: OperationalStatus | "";
  incident_severity: SeverityTier | "";
  incident_status: IncidentStatus | "";
  include_resolved: boolean;
}

export interface MapLayerVisibility {
  zones: boolean;
  pipelines: boolean;
  sensors: boolean;
  valves: boolean;
  gateways: boolean;
  incidents: boolean;
  detections: boolean;
}

export type MapSelection =
  | { kind: "asset"; feature: AssetFeature }
  | { kind: "pipeline"; feature: PipelineFeature }
  | { kind: "incident"; feature: IncidentFeature }
  | { kind: "detection"; feature: DetectionFeature }
  | { kind: "zone"; feature: ZoneFeature }
  | null;

File: plateform/backend/app/schemas/map.py
from datetime import datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.assets import AssetType, OperationalStatus
from app.schemas.incidents import Classification, IncidentStatus, SeverityTier
from app.schemas.telemetry import FreshnessState


class MapFeatureType(str, Enum):
    asset = "asset"
    pipeline = "pipeline"
    incident = "incident"
    zone = "zone"


class GeoJSONGeometry(BaseModel):
    type: Literal["Point", "LineString", "MultiPolygon"]
    coordinates: Any


class ZoneFeatureProperties(BaseModel):
    code: str
    name: str
    country: str
    region: str
    asset_count: int = Field(ge=0)
    active_incident_count: int = Field(ge=0)


class PipelineFeatureProperties(BaseModel):
    external_id: str
    name: str
    status: str
    criticality: int
    zone: str
    population_served: int
    active_incident_count: int = Field(ge=0)
    connected_asset_count: int = Field(ge=0)


class AssetFeatureProperties(BaseModel):
    external_id: str
    name: str
    asset_type: AssetType
    operational_status: OperationalStatus
    health_score: float | None = None
    zone: str
    zone_id: str | None = None
    location_label: str | None = None
    has_cellular_identity: bool = False
    device_msisdn_masked: str | None = None
    pipeline_segment: str | None = None
    active_incident_count: int = Field(ge=0)
    last_seen_at: datetime | None = None
    latest_reading_at: datetime | None = None
    telemetry_freshness: FreshnessState | None = None
    latest_packet_loss_pct: float | None = None
    latest_signal_strength_dbm: int | None = None
    latest_battery_pct: float | None = None


class IncidentFeatureProperties(BaseModel):
    incident_number: str
    title: str
    severity: SeverityTier
    classification: Classification
    status: IncidentStatus
    detected_at: datetime
    related_asset_id: str | None = None
    related_segment_id: str | None = None
    summary: str | None = None


class DetectionFeatureProperties(BaseModel):
    detection_number: str
    priority: str
    status: str
    anomaly_score: float
    trigger_reason: str
    rule_code: str
    sensor_id: str
    zone: str
    detected_at: datetime
    data_mode: str = "simulated"


class MapFeature(BaseModel):
    type: Literal["Feature"] = "Feature"
    id: str
    geometry: GeoJSONGeometry
    properties: (
        ZoneFeatureProperties
        | PipelineFeatureProperties
        | AssetFeatureProperties
        | IncidentFeatureProperties
        | DetectionFeatureProperties
    )


class FeatureCollection(BaseModel):
    type: Literal["FeatureCollection"] = "FeatureCollection"
    features: list[MapFeature]


class MapSummary(BaseModel):
    visible_zones: int = Field(ge=0)
    visible_pipelines: int = Field(ge=0)
    visible_assets: int = Field(ge=0)
    online: int = Field(ge=0)
    degraded: int = Field(ge=0)
    offline: int = Field(ge=0)
    active_incidents: int = Field(ge=0)
    critical_incidents: int = Field(ge=0)
    last_data_update: datetime | None = None
    data_mode: Literal["seeded_demo"] = "seeded_demo"


class NearbyItem(BaseModel):
    feature_type: Literal["asset", "pipeline", "incident"]
    id: str
    name: str
    distance_m: float = Field(ge=0)
    geometry: GeoJSONGeometry
    properties: AssetFeatureProperties | PipelineFeatureProperties | IncidentFeatureProperties


class NearbyResponse(BaseModel):
    origin: dict[str, float]
    radius_m: float
    items: list[NearbyItem]
    total: int = Field(ge=0)
